src/Microsoft.ML.FastTree/TreeTrainersCatalog.cs

```diff
@@ -24,20 +24,31 @@ public static class TreeExtensions
         /// <param name="numLeaves">The maximum number of leaves per decision tree.</param>
         /// <param name="minDatapointsInLeaves">The minimal number of datapoints allowed in a leaf of a regression tree, out of the subsampled data.</param>
         /// <param name="learningRate">The learning rate.</param>
-        /// <param name="advancedSettings">Algorithm advanced settings.</param>
         public static FastTreeRegressionTrainer FastTree(this RegressionContext.RegressionTrainers ctx,
             string labelColumn = DefaultColumnNames.Label,
             string featureColumn = DefaultColumnNames.Features,
             string weights = null,
             int numLeaves = Defaults.NumLeaves,
             int numTrees = Defaults.NumTrees,
             int minDatapointsInLeaves = Defaults.MinDocumentsInLeaves,
-            double learningRate = Defaults.LearningRates,
-            Action<FastTreeRegressionTrainer.Arguments> advancedSettings = null)
+            double learningRate = Defaults.LearningRates)
         {
             Contracts.CheckValue(ctx, nameof(ctx));
             var env = CatalogUtils.GetEnvironment(ctx);
-            return new FastTreeRegressionTrainer(env, labelColumn, featureColumn, weights, numLeaves, numTrees, minDatapointsInLeaves, learningRate, advancedSettings);
+            return new FastTreeRegressionTrainer(env, labelColumn, featureColumn, weights, numLeaves, numTrees, minDatapointsInLeaves, learningRate);
```

Comment:
>FastTreeRegressionTrainer(env, labelColumn, featureColumn, weights, numLeaves, numTrees, minDatapointsInLeaves, learningRate); [](start = 23, length = 126)

So, this is kind of what I mean. While these two extension methods are good, I might have preferred that we just have a *single* constructor (always taking `Options`), and we just keep the flat constructor as simple as possible (e.g., take only the `Options` class, at least publicly). If there are "N" extension methods on this (which we consider the premier public surface!) there don't have to be a corresponding "N" constructors. This could have easily (and indeed more cleanly) acheived with exactly one constructor (again possibly internal) on the underlying object that just took an `Options` object, simply. As it is we're making the underlying object considerably more complex, I think, than it really has to be.

Reply:
Created issue #2100 to  track this. 

For now I have made the constructors internal.

---
In reply to: [246172562](https://github.com/dotnet/machinelearning/pull/2047#discussion_r246172562) [](ancestors = 246172562)